packages/vite/src/node/plugins/resolve.ts

```diff
@@ -395,7 +399,12 @@ export function resolvePlugin(resolveOptions: InternalResolveOptions): Plugin {
         // externalize if building for SSR, otherwise redirect to empty module
         if (isBuiltin(id)) {
           if (ssr) {
-            if (ssrNoExternal === true) {
+            if (
+              ssrNoExternal === true &&
+              // if both noExternal and external are true, noExternal will take the higher priority and bundle it.
+              // only if the id is explicitly listed in external, we will externalize it and skip this error.
+              (ssrExternal === true || !ssrExternal?.includes(id))
```

Comment:
I'm confused, shouldn't this be `ssrExternal === false`?
```suggestion
              (ssrExternal === false || !ssrExternal?.includes(id))
```

Reply:
We don't support `false` for `ssr.external`, only `string[] | true`. It's `true` because (as explained in the comment above), if `noExternal` and `external` are both true, `noExternal` will take priority (meaning we would bundle this builtin but we can't), and so it would enter this block and emit the error.

The behaviour if both are set to `true` is also documented at https://main.vitejs.dev/config/ssr-options.html#ssr-noexternal

Reply:
Ah, I see, thanks for re-explaining it to me!